### PR TITLE
Wait for new master when failing shard

### PR DESCRIPTION
**Pull request by jasontedor**
This commit handles the situation when we are failing a shard and either
no master is known, or the known master left while failing the shard. We
handle this situation by waiting for a new master to be reelected, and
then sending the shard failed request to the new master.

Relates #14252

### CODE REVIEW

**Comment by jasontedor** [comment]
Note that this pull request contains refactorings from #15735 and #15736 that I extracted into separate pull requests.

**Comment by jasontedor** [comment]
@bleskes This is ready for review now.

**Comment by bleskes** [comment]
Discussed with @jasontedor . Decided to try folding all *channel* related retries (connection, master loss and any unknown exception) into the ShardStateAction , as opposed to dealing with that in TransportReplicationAction.

**Comment by jasontedor** [comment]
@bleskes I pushed 6ecedc1c44d5df9327567308e5aceb214acee1e6 to implement what we have discussed.

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 1]
```diff
@@ -555,7 +556,24 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
...
-                    transport.handleResponse(shardFailedRequest.requestId, TransportResponse.Empty.INSTANCE);
+                    if (randomBoolean()) {
+                        // simulate master left and test that the shard failure is retried
+                        int numberOfRetries = randomIntBetween(1, 4);
+                        CapturingTransport.CapturedRequest currentRequest = shardFailedRequest;
+                        for (int retryNumber = 0; retryNumber < numberOfRetries; retryNumber++) {
+                            // force a new cluster state to simulate a new master having been elected
+                            clusterService.setState(ClusterState.builder(clusterService.state()));
+                            transport.handleResponse(currentRequest.requestId, new NotMasterException("shard-failed-test"));
+                            CapturingTransport.CapturedRequest[] retryRequests = transport.capturedRequests();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: we should probably have a  clearAndGetCapturedRequests . We use this often.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -882,15 +882,33 @@ public void handleException(TransportException exp) {
...
-                                logger.warn("{} failed to perform {} on node {}", exp, shardId, transportReplicaAction, node);
-                                shardStateAction.shardFailed(clusterService.state(), shard, indexUUID, "failed to perform " + transportReplicaAction + " on replica on node " + node, exp, shardFailedTimeout, new ReplicationFailedShardStateListener(nodeId, exp));
+                                String message = String.format(Locale.ROOT, "failed to perform %s on replica on node %s", transportReplicaAction, node);
+                                logger.warn("{} {}", exp, shardId, message);
+                                shardStateAction.shardFailed(
+                                    shard,
+                                    indexUUID,
+                                    message,
+                                    exp,
+                                    shardFailedTimeout,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: do we still need this timeout mechanism? if we do, does it better belong in the shard state action?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 3]
```diff
@@ -64,51 +68,94 @@
...
+                    public void handleResponse(TransportResponse.Empty response) {
+                        listener.onSuccess();
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (exp instanceof ReceiveTimeoutTransportException) {
+                            logger.trace("{} timeout sending shard failure [{}] to master [{}]", exp, shardRoutingEntry.getShardRouting().getId(), shardRoutingEntry.failure, masterNode);
+                            handleTimeout(shardRoutingEntry, observer, options, listener);
+                        } else if (exp.getCause() instanceof NotMasterException) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: I think we want to deal with a disconnect here too?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: Also, "FailedToCommitClusterStateException" should be here too. It does bug me that this exception leaks from the master service. We just step down on the spot and say "not a master". But that's different change.

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 4]
```diff
@@ -555,7 +556,24 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
                     // the shard the request was sent to and the shard to be failed should be the same
                     assertEquals(shardRoutingEntry.getShardRouting(), routing);
                     failures.add(shardFailedRequest);
-                    transport.handleResponse(shardFailedRequest.requestId, TransportResponse.Empty.INSTANCE);
+                    if (randomBoolean()) {
+                        // simulate master left and test that the shard failure is retried
+                        int numberOfRetries = randomIntBetween(1, 4);
+                        CapturingTransport.CapturedRequest currentRequest = shardFailedRequest;
+                        for (int retryNumber = 0; retryNumber < numberOfRetries; retryNumber++) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: nice!

**Comment by bleskes** [comment]
I think this looks great and is the right way to go. Left some comments. I'm doubtful if we need the timeout mechanism - it just adds complexity. Let's talk this one over.

Also - can we a open follow up issues to deal with shard started and remove the retry logics from IndicesClusterService ? (will make that class simpler).

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java (hunk 1 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java]: I opened #15897.

**Comment by jasontedor** [comment]
> Also - can we a open follow up issues to deal with shard started

I opened #15895.

> and remove the retry logics from IndicesClusterService ? (will make that class simpler).

I opened #15896.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I removed it in 48b5897215ab6a4ef1f867e57b414c87a517243e.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: Dealing with disconnects was added in 48b5897215ab6a4ef1f867e57b414c87a517243e and handling `FailedtoCommitClusterStateException` was added in be9f8858dff3a726157d760da87d41435d56c1be.

**Comment by jasontedor** [comment]
@bleskes I think this is ready for a final another review round.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 5]
```diff
@@ -92,8 +92,6 @@
  */
 public abstract class TransportReplicationAction<Request extends ReplicationRequest, ReplicaRequest extends ReplicationRequest, Response extends ReplicationResponse> extends TransportAction<Request, Response> {
 
-    public static final String SHARD_FAILURE_TIMEOUT = "action.support.replication.shard.failure_timeout";
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Thank you.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 6]
```diff
@@ -42,73 +44,114 @@
...
-                }
-            });
+    private static Set<Class<?>> MASTER_CHANNEL_EXCEPTIONS =
+        new HashSet<>(Arrays.asList(
+            NotMasterException.class,
+            NodeDisconnectedException.class,
+            Discovery.FailedToCommitClusterStateException.class
+        ));
+    private static boolean isMasterChannelException(Throwable cause) {
+        return MASTER_CHANNEL_EXCEPTIONS.contains(cause.getClass());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: don't we need to unwrap? If so, I think we need to strengthen our IT tests here..

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 7]
```diff
@@ -334,10 +377,7 @@ public String toString() {
         default void onSuccess() {
         }
 
-        default void onShardFailedNoMaster() {
-        }
-
-        default void onShardFailedFailure(final DiscoveryNode master, final TransportException e) {
+        default void onShardFailedFailure(final Exception e) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we add some java docs concerning what exceptions are bubbled here?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 8]
```diff
@@ -50,11 +58,37 @@
 public class ShardStateActionTests extends ESTestCase {
     private static ThreadPool THREAD_POOL;
 
-    private ShardStateAction shardStateAction;
+    private AtomicBoolean timeout;
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: this doesn't seem to be used?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 6 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: The cause is unwrapped in [`ShardStateAction#sendShardFailed`](https://github.com/jasontedor/elasticsearch/blob/be9f8858dff3a726157d760da87d41435d56c1be/core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java#L114).

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: @bleskes The cause is unwrapped in [`ShardStateAction#sendShardFailed`](https://github.com/jasontedor/elasticsearch/blob/be9f8858dff3a726157d760da87d41435d56c1be/core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java#L114).

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 7 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: @bleskes I think we can say what exceptions are *not* bubbled up, but not what exceptions are?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 8 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: @bleskes Removed in 70f02d24c4cf20542bff97315ae4b05444dac29e.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 7 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: @bleskes Added Javadocs in 70f02d24c4cf20542bff97315ae4b05444dac29e.

**Comment by jasontedor** [comment]
@bleskes I pushed more commits.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 9]
```diff
@@ -42,73 +44,114 @@
...
+    private static boolean isMasterChannelException(Throwable cause) {
+        return MASTER_CHANNEL_EXCEPTIONS.contains(cause.getClass());
+    }
+
+    // visible for testing
+    protected void waitForNewMasterAndRetry(ClusterStateObserver observer, ShardRoutingEntry shardRoutingEntry, Listener listener) {
+        observer.waitForNextChange(new ClusterStateObserver.Listener() {
+            @Override
+            public void onNewClusterState(ClusterState state) {
+                sendShardFailed(observer, shardRoutingEntry, listener);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: can we add a trace log here?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java [hunk 10]
```diff
@@ -42,73 +44,114 @@
...
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.error("{} node closed while handling failed shard [{}]", shardRoutingEntry.failure, shardRoutingEntry.getShardRouting().getId(), shardRoutingEntry.getShardRouting());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: doubting if this should be debug. Shutdown is OK.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 11]
```diff
@@ -84,94 +119,105 @@ public static void stopThreadPool() {
...
         final String index = "test";
 
         clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
 
         String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
 
-        AtomicBoolean failure = new AtomicBoolean();
-        assert !failure.get();
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicBoolean noMaster = new AtomicBoolean();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: I'm not sure this noMaster has added value, it basically indicates that our test hook was in place, but we also get it from retried.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 12]
```diff
@@ -84,94 +119,105 @@ public static void stopThreadPool() {
...
+        AtomicBoolean retried = new AtomicBoolean();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: can we make this an AtomicLong and also test that the number of times we retried is what we expect?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 13]
```diff
@@ -84,94 +119,105 @@ public static void stopThreadPool() {
...
+                exception.set(e);
+                latch.countDown();
             }
         });
 
         final CapturingTransport.CapturedRequest[] capturedRequests = transport.getCapturedRequestsAndClear();
         assertThat(capturedRequests.length, equalTo(1));
-        assert !failure.get();
-        transport.handleResponse(capturedRequests[0].requestId, new TransportException("simulated"));
+        assertFalse(success.get());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: don't we want to assert the retries here and at the end of the test as well?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 14]
```diff
@@ -84,94 +119,105 @@ public static void stopThreadPool() {
...
+        shardStateAction.shardFailed(getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
             @Override
-            public void onShardFailedFailure(DiscoveryNode master, TransportException e) {
-                if (e instanceof ReceiveTimeoutTransportException) {
-                    assertFalse(progress.get());
-                    timedOut.set(true);
-                }
-                latch.countDown();
+            public void onShardFailedFailure(Exception e) {
+                failure.set(true);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: can we make sure success is not called?

**Comment by bleskes** [comment]
Thanks @jasontedor . Looks great. I left some comments w.r.t testing. A couple of things I miss in that area:

1) A simple test for success - without a retry/exceptions.
2) Validation of the message sent to the master - we learned from bitter experience that we should test that ... :)

**Comment by jasontedor** [comment]
> Validation of the message sent to the master - we learned from bitter experience that we should test that 

@bleskes Can you elaborate what you mean? I have some commits locally I was planning for another PR (checking if the shard is in the routing table, checking if the shard is a replica not being failed by the primary, checking if the shard is a primary not being failed by the owning node) but want to be sure we're thinking the same thing.

**Comment by bleskes** [comment]
I meant that we should check the message is the right class and that it contains the shard routing we gave it.

**Comment by jasontedor** [comment]
> I meant that we should check the message is the right class and that it contains the shard routing we gave it.

Oh, in the tests. ACK.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 12 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: @bleskes I'm concerned this is going to make already-complex tests too complex. What do you think?

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 13 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: Addressed in efb142613ffc38b30db81cf7b01e5c2d24f1baff.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 11 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: Addressed in efb142613ffc38b30db81cf7b01e5c2d24f1baff.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 14 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: Addressed in efb142613ffc38b30db81cf7b01e5c2d24f1baff.

**Comment by jasontedor** [comment]
> 1) A simple test for success - without a retry/exceptions.
2) Validation of the message sent to the master - we learned from bitter experience that we should test that ... :)

@bleskes Pushed a new test in efb142613ffc38b30db81cf7b01e5c2d24f1baff.

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 10 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: `WARN` then?

File: core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java (hunk 9 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java]: I pushed fe39d11c55c5fea0518bf0c38c87cd15343416e7.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 12 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: I added a retry loop in 7eefcbbeed4ca06aa8e5b4ff853dbb0d8d6cfde1. :)

**Comment by jasontedor** [comment]
@bleskes I think this pull request is ready for another retry loop. ;)

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java [hunk 15]
```diff
@@ -84,94 +123,149 @@ public static void stopThreadPool() {
...
+        AtomicInteger retries = new AtomicInteger();
+        AtomicBoolean success = new AtomicBoolean();
+
+        setUpMasterRetryVerification(1, retries, latch, requestId -> {});
 
-        shardStateAction.shardFailed(clusterService.state(), getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
+        shardStateAction.shardFailed(getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
             @Override
-            public void onShardFailedNoMaster() {
+            public void onSuccess() {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: can we make sure onShardFailedFailure is not called?

**Comment by bleskes** [comment]
LGTM. Left one little comment that doesn't need another cycle. Thanks @jasontedor !

**Comment by jasontedor** [comment]
Thanks for another very thorough and helpful review @bleskes.

File: core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java (hunk 15 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java]: Addressed in 386d2ab86a33fa83cfa770f0d37542572136736f.

**Pull request merged**